src/proof/types.rs

```diff
@@ -1,56 +1,91 @@
 use super::*;
-use rlp::{RlpStream};
-use ethereum_types::{Bloom, H160, H256, H64, U256};
+use ethereum_types::{Address, Bloom, H256, H64, U256};
+use rlp::RlpStream;
+use rlp_derive::{RlpDecodable, RlpEncodable};
 
-#[derive(Debug, Clone)]
+#[derive(Debug, Clone, RlpEncodable, RlpDecodable)]
 pub struct BlockHeader {
-	pub parent_hash: H256,
-	pub uncles_hash: H256,
-	pub author: H160,
-	pub state_root: H256,
-	pub transactions_root: H256,
-	pub receipts_root: H256,
-	pub log_bloom: Bloom,
-	pub difficulty: U256,
-	pub number: u64,
-	pub gas_limit: U256,
-	pub gas_used: U256,
-	pub timestamp: u64,
-	pub extra_data: Vec<u8>,
-	pub mix_hash: H256,
-	pub nonce: H64,
+    pub parent_hash: H256,
+    pub uncles_hash: H256,
+    pub author: Address,
+    pub state_root: H256,
+    pub transactions_root: H256,
+    pub receipts_root: H256,
+    pub log_bloom: Bloom,
+    pub difficulty: U256,
+    pub number: U256,
+    pub gas_limit: u64,
+    pub gas_used: u64,
+    pub timestamp: u64,
+    pub extra_data: String,
```

Comment:
Note: here it is important to use `String` and not just `Vec<u8>` since the `Vec<T>` will be treated as an RLP List, not just normal raw bytes, but we should use another thing other than a UTF8 String since maybe this extra data is not Utf8 safe.